py/picca/pk1d/postproc_pk1d.py

```diff
@@ -862,12 +818,370 @@ def compute_average_pk_redshift(
+    nbins_k,
+    izbin,
+    p1d_weights,
+    covariance_weights,
+    p1d_groups,
+):
+    """Compute the covariance of a set of 1D power spectra.
+
+    Arguments
+    ---------
+    weight_method: str,
+    Method to weight the data.
+
+    mean_p1d_table (array-like):
+    Table of mean 1D power spectra, with column 'meanPk'.
+
+    nbins_k (int):
+    Number of k bins.
+
+    izbin (int):
+    Current redshift bin being considered.
+
+    p1d_weights  (array-like):
+    Weights associated with p1d pixels for all subforest, used in the calculation of covariance.
+
+    covariance_weights (array-like):
+    Weights for all subforest used inside the main covariance sum.
+
+    p1d_groups (array-like):
+    Individual p1d pixels grouped in the same wavenumber binning for all subforest
+
+    Return
+    ------
+    covariance_array (array-like):
+    Array of covariance coefficients.
+    """
+
+    if len(p1d_groups) == 0:
+        return np.full(nbins_k * nbins_k, np.nan)
+
+    mean_pk = mean_p1d_table["meanPk"][(nbins_k * izbin) : (nbins_k * (izbin + 1))]
+    mean_pk_product = np.outer(mean_pk, mean_pk)
+
+    sum_p1d_weights = np.nansum(p1d_weights, axis=0)
+    weights_sum_product = np.outer(sum_p1d_weights, sum_p1d_weights)
+
+    p1d_groups_product_sum = np.zeros((nbins_k, nbins_k))
+    covariance_weights_product_sum = np.zeros((nbins_k, nbins_k))
+    weights_product_sum = np.zeros((nbins_k, nbins_k))
+
+    for i, p1d_group in enumerate(p1d_groups):
+        p1d_groups_product_sum = np.nansum(
+            [p1d_groups_product_sum, np.outer(p1d_group, p1d_group)], axis=0
+        )
+        covariance_weights_product_sum = np.nansum(
+            [
+                covariance_weights_product_sum,
+                np.outer(covariance_weights[i], covariance_weights[i]),
+            ],
+            axis=0,
+        )
+        weights_product_sum = np.nansum(
+            [weights_product_sum, np.outer(p1d_weights[i], p1d_weights[i])], axis=0
+        )
+
+    del p1d_groups, covariance_weights, p1d_weights
+
+    covariance_matrix = (weights_product_sum / weights_sum_product) * (
+        (p1d_groups_product_sum / covariance_weights_product_sum) - mean_pk_product
+    )
+
+    # For fit_snr method, due to the SNR fitting scheme used for weighting,
+    # the diagonal of the weigthed sample covariance matrix is not equal
+    # to the error in mean P1D. This is tested on Ohio mocks and data.
+    # We choose to renormalize the whole covariance matrix.
+    if weight_method == "fit_snr":
+        std_pk = mean_p1d_table["errorPk"][(nbins_k * izbin) : (nbins_k * (izbin + 1))]
+        covariance_diag = np.diag(covariance_matrix)
+        covariance_matrix = (
+            covariance_matrix
+            * np.outer(std_pk, std_pk)
+            / np.sqrt(np.outer(covariance_diag, covariance_diag))
+        )
+
+    covariance_array = np.ravel(covariance_matrix)
+
+    return covariance_array
+
+
+def compute_cov_not_vectorized(
```

Comment:
this is not only not vectorized, but the actual complete old method, i.e. not only slower, but also doing slightly different things potentially by actually summing up each individual mode every time. Maybe clarify that in the docstring

Reply:
That is what we are doing on the new method too, no ?
I have added a comment on this function

Reply:
yes, but in the old one the approach is split in a completely different way by looping over every mode while in the new we're precomputing intermediate summaries that are then reused multiple times, and in addition do everything in a vectorized way...